physics/kepler_orbit_body.hpp

```diff
@@ -157,94 +213,478 @@ KeplerOrbit<Frame>::KeplerOrbit(
+  Length const& ℓ = *elements_at_epoch_.semilatus_rectum;
+  SpecificEnergy const& ε = *elements_at_epoch_.specific_energy;
+  KeplerianElements<Frame> elements = elements_at_epoch_;
+  elements.true_anomaly.reset();
+  elements.mean_anomaly.reset();
+  elements.hyperbolic_mean_anomaly.reset();
+  if (e < 1) {
+    elements.mean_anomaly = *elements_at_epoch_.mean_anomaly +
+                            *elements_at_epoch_.mean_motion * (t - epoch_);
+  } else if (e == 1) {
     // Parabolic case.
     LOG(FATAL) << "not yet implemented";
-    base::noreturn();
   } else {
-    // Hyperbolic case.
-    LOG(FATAL) << "not yet implemented";
-    base::noreturn();
+    elements.hyperbolic_mean_anomaly =
+        *elements_at_epoch_.hyperbolic_mean_anomaly +
+        *elements_at_epoch_.hyperbolic_mean_motion * (t - epoch_);
   }
+  CompleteAnomalies(elements);
+  Angle const& ν = *elements.true_anomaly;
+  struct OrbitPlane;
+  Rotation<OrbitPlane, Frame> const from_orbit_plane(
+      Ω, i, ω,
+      EulerAngles::ZXZ,
+      DefinesFrame<OrbitPlane>{});
+  Length const r = ℓ / (1 + e * Cos(ν));
+  Displacement<Frame> const displacement =
+      r * from_orbit_plane(Vector<double, OrbitPlane>({Cos(ν), Sin(ν), 0}));
+  // Flight path angle.
+  Angle const φ = ArcTan(e * Sin(ν), 1 + e * Cos(ν));
+  // The norm comes from the vis-viva equation.
+  Velocity<Frame> const velocity =
+      Sqrt(2 * (ε + μ / r)) *
+      from_orbit_plane(Vector<double, OrbitPlane>(
+          {-Sin(ν - φ), Cos(ν - φ), 0}));
+  return {displacement, velocity};
 }
 
 template<typename Frame>
 KeplerianElements<Frame> const& KeplerOrbit<Frame>::elements_at_epoch() const {
   return elements_at_epoch_;
 }
 
+template<typename Frame>
+void KeplerOrbit<Frame>::CompleteConicParametersByCategory(
+    KeplerianElements<Frame>& elements,
+    GravitationalParameter const& μ) {
+  auto& eccentricity = elements.eccentricity;
+  auto& asymptotic_true_anomaly = elements.asymptotic_true_anomaly;
+  auto& turning_angle = elements.turning_angle;
+  auto& semimajor_axis = elements.semimajor_axis;
+  auto& specific_energy = elements.specific_energy;
+  auto& characteristic_energy = elements.characteristic_energy;
+  auto& mean_motion = elements.mean_motion;
+  auto& period = elements.period;
+  auto& hyperbolic_mean_motion = elements.hyperbolic_mean_motion;
+  auto& hyperbolic_excess_velocity = elements.hyperbolic_excess_velocity;
+  auto& semiminor_axis = elements.semiminor_axis;
+  auto& impact_parameter = elements.impact_parameter;
+  auto& semilatus_rectum = elements.semilatus_rectum;
+  auto& specific_angular_momentum = elements.specific_angular_momentum;
+  auto& periapsis_distance = elements.periapsis_distance;
+  auto& apoapsis_distance = elements.apoapsis_distance;
+  int const eccentricity_specifications = eccentricity.has_value() +
+                                          asymptotic_true_anomaly.has_value() +
+                                          turning_angle.has_value();
```

Comment:
Again, when reading this code I wonder if we could have some kind of a oneof class to help enforce these invariants.